pkg/util/configz/configz.go

```diff
@@ -0,0 +1,86 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package configz
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"sync"
+)
+
+var (
+	configsGuard sync.RWMutex
+	configs      = map[string]*Config{}
+)
+
+func InstallHandler(m mux) {
+	m.Handle("/configz", http.HandlerFunc(handle))
+}
+
+type Config struct {
+	val interface{}
+}
+
+func New(name string) (*Config, error) {
+	configsGuard.Lock()
+	defer configsGuard.Unlock()
+	if _, found := configs[name]; found {
+		return nil, fmt.Errorf("register config %q twice", name)
+	}
+	newConfig := Config{}
+	configs[name] = &newConfig
+	return &newConfig, nil
+}
+
+func Delete(name string) {
+	configsGuard.Lock()
+	defer configsGuard.Unlock()
+	delete(configs, name)
+}
+
+func (v *Config) Set(val interface{}) {
+	configsGuard.Lock()
+	defer configsGuard.Unlock()
+	v.val = val
+}
+
+func (v *Config) MarshalJSON() ([]byte, error) {
+	return json.Marshal(v.val)
```

Comment:
The configsGuard rwlock guards the configs map from concurrent modification while we traverse. We are holding a readlock here already and by the time we get here, we're already traversing the map. I've assumed that the config objects that are added are never modified and there are no locks corresponding to those objects I think that's a pretty safe assumption.